### PR TITLE
BEEFY: generate historical proofs

### PR DESCRIPTION
**Pull request by svyatonik**
Polkadot companion: https://github.com/paritytech/polkadot/pull/5334

This PR adds new MMR RPC method - `mmr_generateHistoricalProof`. It is quite similar to the existing `mmr_generateProof` method, but it has additional argument - number of MMR leaves. It may be used to generate proof using historical MMR state even using latest block state, where there are more leaves.

Where it is useful:
1) BEEFY light client needs to import explicit commitment + leaf + leaf proof for blocks that are changing validator set. If the chain has progressed since such block, the state of validator-set-change-block may have already been discarded (unless you're using archive node, which we want to avoid) and the call of `mmr_generateProof(use-state-at-validator-set-change-block)` will obviously fail. We may use `mmr_generateHistoricalProof(use-state-at-best-block, use-mmr-state-at-validator-set-change-block)` instead - this will generate us required proof;
2) if block `B` is known to the light client, then we can't simply generate some MMR-related proof using MMR state at block `B+10`. So e.g. to prove some storage (read: deliver message), we need to generate proof using MMR state at block `B`. If state is discarded, we'll need to import `B+10` first => extra bridge cost.

The `leaves_count` argument can easily be computed by the client by assuming that there's exactly one leaf per every block starting from some beginning block (where MMR pallet has been deployed). We may read number of MMR leaves at the `best_block` and then compute `leaves_count_at_historical_block` as `number_of_historical_block - (best_block_number - leaves_count_at_best_block)`.

### CODE REVIEW

**Review by acatangiu — APPROVED** [review]

File: primitives/merkle-mountain-range/src/lib.rs [hunk 1]
```diff
@@ -399,6 +399,17 @@ sp_api::decl_runtime_apis! {
...
+		/// Generate a historical MMR proof for a leaf with given index.
+		///
+		/// This function may be used to generate proofs using MMR state at some earlier
+		/// point. To do that, you shall provide historical value of the `leaves_count`.
+		/// The proof generated by this function may be verified by clients that only have
+		/// some old MMR root (corresponding to the `leaves_count`).
+		fn generate_historical_proof(
+			leaf_index: LeafIndex,
+			leaves_count: LeafIndex,
+		) -> Result<(EncodableOpaqueLeaf, Proof<Hash>), Error>;
```
Comment (acatangiu) [inline primitives/merkle-mountain-range/src/lib.rs:411]: Would it make sense to directly add this with support for multiple leaves at once? (see https://github.com/paritytech/substrate/pull/10635)

Would the light client benefit for generating historical proofs in a batch?

Comment (svyatonik) [reply primitives/merkle-mountain-range/src/lib.rs:411]: Yeah - I think it makes sense to add support for proving several leaves, even though we'll only be using this API to prove single leaf now.

Comment (svyatonik) [reply primitives/merkle-mountain-range/src/lib.rs:411]: Then let's probably defer this PR until #10635 is merged to reuse primitives from there && etc

File: frame/merkle-mountain-range/src/lib.rs [hunk 2]
```diff
@@ -265,7 +265,23 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+
+	/// Generate a MMR proof for the given `leaf_index`.
+	///
+	/// This method is not touching any runtime storage keys, as all it needs is coming
+	/// either from arguments, or from the offchain storage.
+	///
+	/// Note this method can only be used from an off-chain context
+	/// (Offchain Worker or Runtime API call), since it requires
+	/// all the leaves to be present.
+	/// It may return an error or panic if used incorrectly.
```
Comment (tomusdrw) [inline frame/merkle-mountain-range/src/lib.rs:279]: IMHO it makes sense to add an explicit error in case `leaves_count > Self::mmr_leaves()`

**Pull request closed without merging**